server/src/controllers/aplus.ts

```diff
@@ -0,0 +1,217 @@
+import {validateAttainmentPath} from './utils/attainment';
+import {APLUS_API_URL} from '../configs/environment';
+import AplusGradeSource from '../database/models/aplusGradeSource';
+import {ApiError} from '../types';
+
+/**
+ * Responds with AplusExerciseData
+ *
+ * @throws ApiError(400|502)
+ */
+export const fetchAplusExerciseData = async (
+  req: Request,
+  res: Response
+): Promise<void> => {
+  const aplusCourseId = validateAplusCourseId(req.params.aplusCourseId);
+
+  const exercisesRes = await fetchFromAplus<{
+    results: {
+      id: number;
+      display_name: string;
+      exercises: {
+        difficulty: string;
+      }[];
+    }[];
+  }>(`${APLUS_API_URL}/courses/${aplusCourseId}/exercises?format=json`);
+
+  // There doesn't appear to be a better way to get difficulties
+  const difficulties = new Set<string>();
+  for (const result of exercisesRes.data.results) {
+    for (const exercise of result.exercises) {
+      if (exercise.difficulty) {
+        difficulties.add(exercise.difficulty);
+      }
+    }
+  }
+
+  const exerciseData: AplusExerciseData = {
+    modules: exercisesRes.data.results.map(result => ({
+      id: result.id,
+      name: result.display_name,
+    })),
+    difficulties: Array.from(difficulties),
+  };
+
+  res.json(exerciseData);
+};
+
+/** @throws ApiError(400|404|409) */
+export const addAplusGradeSources = async (
+  req: TypedRequestBody<typeof NewAplusGradeSourceArraySchema>,
+  res: Response
+): Promise<void> => {
+  const newGradeSources: NewAplusGradeSourceData[] = req.body;
+  for (const newGradeSource of newGradeSources) {
+    await validateAttainmentPath(
+      req.params.courseId,
+      String(newGradeSource.attainmentId)
+    );
+  }
+
+  await AplusGradeSource.bulkCreate(newGradeSources);
+
+  res.sendStatus(HttpCode.Created);
+};
+
+/**
+ * Responds with NewGrade[]
+ *
+ * @throws ApiError(400|404|409|422|502)
+ */
+export const fetchAplusGrades = async (
+  req: Request,
+  res: Response
+): Promise<void> => {
+  let attainmentIds: number[] = [];
+  try {
+    attainmentIds = z
+      .array(IdSchema)
+      .parse(JSON.parse(String(req.query.attainments)));
+  } catch (e) {
+    if (e instanceof Error) {
+      throw new ApiError(e.message, HttpCode.BadRequest);
+    }
+    if (e instanceof ZodError) {
+      throw new ApiError(
+        e.issues.map(issue => issue.message),
+        HttpCode.BadRequest
+      );
+    }
+  }
+
+  const newGrades: NewGrade[] = [];
+  for (const attainmentId of attainmentIds) {
+    const [_, attainment] = await validateAttainmentPath(
+      req.params.courseId,
+      String(attainmentId)
+    );
+
+    // TODO: There can be multiple sources
+    const gradeSource = await AplusGradeSource.findOne({
+      where: {attainmentId: attainment.id},
+    });
+
+    if (!gradeSource) {
+      throw new ApiError(
+        `attainment with ID ${attainment.id} has no A+ grade sources`,
+        HttpCode.UnprocessableEntity
```

Comment:
Shouldn't this be `404 Not Found`?

Reply:
Good question, I'm not sure. The grade source row isn't exactly a requested resource by the API caller, but rather something that needs to exist for the request to be carried out, so based on how 422 has been used in Grades previously, my first intuition was to use that.